tensorflow_probability/python/mcmc/dual_averaging_step_size_adaptation.py

```diff
@@ -0,0 +1,299 @@
+"""DualAveragingStepSizeAdaptation TransitionKernel."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import collections
+
+import numpy as np
+import tensorflow as tf
+
+from tensorflow_probability.python.internal import dtype_util
+from tensorflow_probability.python.mcmc import kernel as kernel_base
+from tensorflow_probability.python.mcmc.internal import util as mcmc_util
+
+
+@mcmc_util.make_innermost_setter
+def _hmc_like_step_size_setter_fn(kernel_results, new_step_size):
+  return kernel_results._replace(
+    accepted_results=kernel_results.accepted_results._replace(
+      step_size=new_step_size
+    )
+  )
+
+
+@mcmc_util.make_innermost_getter
+def _hmc_like_step_size_getter_fn(kernel_results):
+  return kernel_results.accepted_results.step_size
+
+
+@mcmc_util.make_innermost_getter
+def _hmc_like_log_accept_prob_getter_fn(kernel_results):
+  log_accept_ratio = kernel_results.log_accept_ratio
+  return tf.minimum(tf.constant(0.0, log_accept_ratio.dtype), log_accept_ratio)
+
+
+def _reduce_logmeanexp(value, dims, keepdims=False):
+  # This is intentionally numerically imprecise for simplicity. For the purposes
+  # of computing the mean acceptance probability this is more than sufficient.
+  return tf.math.log(
+    tf.reduce_mean(input_tensor=tf.exp(value), axis=dims, keepdims=keepdims)
+  )
+
+
+def _get_differing_dims(a, b):
+  # Get the indices of dimensions where shapes of `a` and `b` differ.
+  # `a` is allowed to have fewer dimensions than `b`.
+  if a.shape.is_fully_defined() and b.shape.is_fully_defined():
+    a_shape = np.array(a.shape.as_list())
+    b_shape = np.array(b.shape.as_list())
+    return np.where(a_shape != b_shape[: len(a_shape)])[0]
+  else:
+    return tf.where(
+      tf.not_equal(tf.shape(input=a), tf.shape(input=b)[: tf.rank(a)])
+    )[:, 0]
+
+
+class DualAveragingStepSizeAdaptationResults(
+  collections.namedtuple(
+    "DualAveragingStepSizeAdaptationResults",
+    "inner_results, target_accept_prob, mu, gamma, t0, kappa, error_sum, log_averaging_step, step, new_step_size",
+  )
+):
+  __slots__ = ()
+
+
+class DualAveragingStepSizeAdaptation(kernel_base.TransitionKernel):
+  def __init__(
+    self,
+    inner_kernel,
+    num_adaptation_steps,
+    target_accept_prob=0.75,
+    gamma=0.05,
+    t0=10.0,
+    kappa=0.75,
+    step_size_setter_fn=_hmc_like_step_size_setter_fn,
+    step_size_getter_fn=_hmc_like_step_size_getter_fn,
+    log_accept_prob_getter_fn=_hmc_like_log_accept_prob_getter_fn,
+    validate_args=False,
+    name=None,
+  ):
+    inner_kernel = mcmc_util.enable_store_parameters_in_results(inner_kernel)
```

Comment:
We'll need to add class and init docstrings in the style of other TransitionKernel objects.